Mathlib/Algebra/Star/StarRingHom.lean

```diff
@@ -0,0 +1,467 @@
+@[coe]
+def toStarRingEquiv {F A B : Type*} [Add A] [Mul A] [Star A] [Add B] [Mul B] [Star B]
+    [EquivLike F A B] [RingEquivClass F A B] [StarRingEquivClass F A B] (f : F) : A  ≃⋆+* B :=
+  { (f : A ≃+* B) with
+    map_star' := map_star f}
+
+/-- Any type satisfying `StarRingEquivClass` can be cast into `StarRingEquiv` via
+`StarRingEquivClass.toStarRingEquiv`. -/
+instance instCoeHead {F A B : Type*} [Add A] [Mul A] [Star A] [Add B] [Mul B] [Star B]
+    [EquivLike F A B] [RingEquivClass F A B] [StarRingEquivClass F A B] : CoeHead F (A ≃⋆+* B) :=
+  ⟨toStarRingEquiv⟩
+
+end StarRingEquivClass
+
+namespace StarRingEquiv
+
+section Basic
+
+variable {A B C : Type*} [Add A] [Add B] [Mul A] [Mul B] [Star A] [Star B] [Add C] [Mul C] [Star C]
+
+instance : EquivLike (A ≃⋆+* B) A B where
+  coe f := f.toFun
+  inv f := f.invFun
+  left_inv f := f.left_inv
+  right_inv f := f.right_inv
+  coe_injective' f g h₁ h₂ := by
+    rcases f with ⟨⟨⟨_, _, _⟩, _⟩, _⟩
+    rcases g with ⟨⟨⟨_, _, _⟩, _⟩, _⟩
+    congr
+
+instance : RingEquivClass (A ≃⋆+* B) A B where
+  map_mul f := f.map_mul'
+  map_add f := f.map_add'
+
+instance : StarRingEquivClass (A ≃⋆+* B) A B where
+  map_star := map_star'
+
+/-- Helper instance for cases where the inference via `EquivLike` is too hard. -/
+instance : FunLike (A ≃⋆+* B) A B where
+  coe f := f.toFun
+  coe_injective' := DFunLike.coe_injective
+
+@[simp]
+theorem toRingEquiv_eq_coe (e : A ≃⋆+* B) : e.toRingEquiv = e :=
+  rfl
+
+@[ext]
+theorem ext {f g : A ≃⋆+* B} (h : ∀ a, f a = g a) : f = g :=
+  DFunLike.ext f g h
+
+theorem ext_iff {f g : A ≃⋆+* B} : f = g ↔ ∀ a, f a = g a :=
+  DFunLike.ext_iff
+
+/-- Star ring equivalences are reflexive. -/
+@[refl]
+def refl : A ≃⋆+* A :=
+  { RingEquiv.refl A with
+    map_star' := fun _ => rfl }
+#align star_alg_equiv.refl StarRingEquiv.refl
+
+instance : Inhabited (A ≃⋆+* A) :=
+  ⟨refl⟩
+
+@[simp]
+theorem coe_refl : ⇑(refl : A ≃⋆+* A) = id :=
+  rfl
+
+/-- Star ring equivalences are symmetric. -/
+@[symm]
+nonrec def symm (e : A ≃⋆+* B) : B ≃⋆+* A :=
+  { e.symm with
+    map_star' := fun b => by
+      simpa only [apply_inv_apply, inv_apply_apply] using
+        congr_arg (inv e) (map_star e (inv e b)).symm }
+
+/-- See Note [custom simps projection] -/
+def Simps.apply (e : A ≃⋆+* B) : A → B := e
+
+/-- See Note [custom simps projection] -/
+def Simps.symm_apply (e : A ≃⋆+* B) : B → A :=
+  e.symm
+
+initialize_simps_projections StarRingEquiv (toFun → apply, invFun → symm_apply)
+
+@[simp]
+theorem invFun_eq_symm {e : A ≃⋆+* B} : EquivLike.inv e = e.symm :=
+  rfl
+
+@[simp]
+theorem symm_symm (e : A ≃⋆+* B) : e.symm.symm = e := by
+  ext
+  rfl
+
+theorem symm_bijective : Function.Bijective (symm : (A ≃⋆+* B) → B ≃⋆+* A) :=
+  Function.bijective_iff_has_inverse.mpr ⟨_, symm_symm, symm_symm⟩
+
+@[simp]
+theorem mk_coe' (e : A ≃⋆+* B) (f h₁ h₂ h₃ h₄ h₅) :
+    (⟨⟨⟨f, e, h₁, h₂⟩, h₃, h₄⟩, h₅⟩ : B ≃⋆+* A) = e.symm :=
+  symm_bijective.injective <| ext fun _ => rfl
```

Comment:
I really don't think this is the theorem we want. It's unwrapping too many structures, and why `symm`?

Reply:
This is _mutatis mutandis_ from  https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Algebra/Star/StarAlgHom.lean#L904

Reply:
Both need to be fixed then. It's likely a consequence of the port.

Reply:
Mathlib3 algbera version: https://github.com/leanprover-community/mathlib/blob/master/src/algebra/star/star_alg_hom.lean#L619
```
@[simp] lemma mk_coe' (e : A ≃⋆ₐ[R] B) (f h₁ h₂ h₃ h₄ h₅ h₆) :
  (⟨f, e, h₁, h₂, h₃, h₄, h₅, h₆⟩ : B ≃⋆ₐ[R] A) = e.symm :=
symm_bijective.injective $ ext $ λ x, rfl
```

Reply:
Perhaps we could just deprecate them? #13267

Reply:
@j-loreaux now #13267 has been queued for merging, I've made the corresponding update to this PR.